File: ide-features-trainer/src/training/learn/LessonProcessor.java
package training.learn;

import com.intellij.openapi.actionSystem.ActionManager;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import org.jdom.Element;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.TestOnly;
import training.commands.Command;
import training.commands.CommandFactory;
import training.commands.ExecutionList;
import training.util.PerformActionUtil;
import training.editor.MouseListenerHolder;
import training.editor.actions.HideProjectTreeAction;
import training.ui.Message;
import training.util.XmlUtil;

import java.util.HashMap;
import java.util.Queue;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.LinkedBlockingQueue;


/**
 * Created by karashevich on 30/01/15.
 */
public class LessonProcessor {

    @TestOnly
    public static ExecutionList getCurrentExecutionList() {
        return currentExecutionList;
    }

    private static ExecutionList currentExecutionList;

    public static void process(final Project project, final Lesson lesson, final Editor editor, @Nullable String target) throws Exception {

        HashMap<String, String> editorParameters = new HashMap<>();

        Queue<Element> elements = new LinkedBlockingQueue<>();
        if (lesson.getScn() == null) {
            System.err.println("Scenario is empty or cannot be read!");
            return;
        }

        final Element root = lesson.getScn().getRoot();

        if (root == null) {
            System.err.println("Scenario is empty or cannot be read!");
            return;
        }

        //getEditor parameters
        getEditorParameters(root, editorParameters);

        //Create queue of Actions
        for (final Element el : root.getChildren()) {
            //if element is MouseBlocked (blocks all mouse events) than add all children inside it.
            if (isMouseBlock(el)) {
                if (el.getChildren() != null) {
                    elements.add(el); //add block element
                    for (Element el1 : el.getChildren()) {
                        if (isCaretBlock(el1)) {
                            if (el1.getChildren() != null) {
                                elements.add(el1); //add block element
                                for (Element el2 : el1.getChildren()) {
                                    elements.add(el2); //add inner elements
                                }
                                elements.add(new Element(Command.CommandType.CARETUNBLOCK.toString())); //add unblock element
                            }
                        } else {
                            elements.add(el1); //add inner elements
                        }
                    }
                    elements.add(new Element(Command.CommandType.MOUSEUNBLOCK.toString())); //add unblock element
                }
            } else if (isCaretBlock(el)) {
                if (el.getChildren() != null) {
                    elements.add(el); //add block element
                    for (Element el1 : el.getChildren()) {
                        elements.add(el1); //add inner elements
                    }
                    elements.add(new Element(Command.CommandType.CARETUNBLOCK.toString())); //add unblock element
                }
            } else {
                elements.add(el);
            }
        }

        MouseListenerHolder mouseListenerHolder = new MouseListenerHolder();

        //Initialize lesson in the editor

        LessonManager lessonManager = LessonManager.getInstance(lesson);
        if (lessonManager != null) {
            lessonManager.initLesson(editor);
        } else {
            throw new Exception("Unable to get LessonManager");
        }

        //Prepare environment before execution
        prepareEnvironment(editor, project, editorParameters);

        //Perform first action, all next perform like a chain reaction
        Command cmd = CommandFactory.buildCommand(elements.peek());
        ExecutionList executionList = new ExecutionList(elements, lesson, project, editor, mouseListenerHolder, target);
        currentExecutionList = executionList;

        cmd.execute(executionList);

    }


    private static void getEditorParameters(Element root, HashMap<String, String> editorParameters) {
        if (root.getAttribute(Lesson.EditorParameters.PROJECT_TREE) != null) {
            editorParameters.put(Lesson.EditorParameters.PROJECT_TREE, root.getAttributeValue(Lesson.EditorParameters.PROJECT_TREE));
        }
    }

    private static void prepareEnvironment(Editor editor, Project project, HashMap<String, String> editorParameters) throws ExecutionException, InterruptedException {
        if (editorParameters.containsKey(Lesson.EditorParameters.PROJECT_TREE)) {
            if (ActionManager.getInstance().getAction(HideProjectTreeAction.actionId) == null) {
                HideProjectTreeAction hideAction = new HideProjectTreeAction();
                ActionManager.getInstance().registerAction(hideAction.getActionId(), hideAction);
            }
            PerformActionUtil.performAction(HideProjectTreeAction.actionId, editor, project);
        }
    }

    private static boolean isMouseBlock(Element el) {
        return el.getName().toUpperCase().equals(Command.CommandType.MOUSEBLOCK.toString());
    }

    private static boolean isCaretBlock(Element el) {
        return el.getName().toUpperCase().equals(Command.CommandType.CARETBLOCK.toString());
    }

    public static String takeDescriptionsOnly(Lesson lesson) {
        StringBuilder sb = new StringBuilder();
        Queue<Element> elements = new LinkedBlockingQueue<>();
        if (lesson.getScn() == null) {
            System.err.println("Scenario is empty or cannot be read!");
            return null;
        }

        final Element root = lesson.getScn().getRoot();

        if (root == null) {
            System.err.println("Scenario is empty or cannot be read!");
            return null;
        }


        //Create queue of Actions
        for (final Element el : root.getChildren()) {
            //if element is MouseBlocked (blocks all mouse events) than add all children inside it.
            if (isMouseBlock(el)) {
                if (el.getChildren() != null) {
                    elements.add(el); //add block element
                    for (Element el1 : el.getChildren()) {
                        if (isCaretBlock(el1)) {
                            if (el1.getChildren() != null) {
                                elements.add(el1); //add block element
                                for (Element el2 : el1.getChildren()) {
                                    elements.add(el2); //add inner elements
                                }
                                elements.add(new Element(Command.CommandType.CARETUNBLOCK.toString())); //add unblock element
                            }
                        } else {
                            elements.add(el1); //add inner elements
                        }
                    }
                    elements.add(new Element(Command.CommandType.MOUSEUNBLOCK.toString())); //add unblock element
                }
            } else if (isCaretBlock(el)) {
                if (el.getChildren() != null) {
                    elements.add(el); //add block element
                    for (Element el1 : el.getChildren()) {
                        elements.add(el1); //add inner elements
                    }
                    elements.add(new Element(Command.CommandType.CARETUNBLOCK.toString())); //add unblock element
                }
            } else {
                elements.add(el);
            }
        }

        while (elements.size() > 0) {
            final Element polledElement = elements.poll();
            if (polledElement.getAttribute("description") != null) {
                String htmlText = polledElement.getAttribute("description").getValue();
                final Message[] messages = XmlUtil.extractAll(new Message[]{new Message(htmlText, Message.MessageType.TEXT_REGULAR)});
                StringBuilder messageString = new StringBuilder();
                for (Message message : messages) {
                    messageString.append(message.getText());
                }
                sb.append(messageString.toString()).append("\n");

            }
        }

        return sb.toString();
    }

}

File: ide-features-trainer/src/training/learn/lesson/LessonProcessor.kt
// Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package training.learn.lesson

import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.project.Project
import org.jdom.Element
import training.commands.Command
import training.commands.CommandFactory
import training.commands.ExecutionList
import java.util.concurrent.LinkedBlockingQueue

object LessonProcessor {

  val LOG = Logger.getInstance(LessonProcessor::class.java)

  @get:org.jetbrains.annotations.TestOnly
  var currentExecutionList: ExecutionList? = null
    private set

  var tasksNumber: Int = 0
    private set
  var currentTaskIndex: Int = 0

  fun process(project: Project, lesson: XmlLesson, editor: Editor?, documentationMode: Boolean = false) {
    val elements = createListOfCommands(lesson.scenario.root)
    val myQueueOfElements = LinkedBlockingQueue(elements)

    //Initialize lesson in the editor
    if (editor != null) {
      LessonManager.instance.initLesson(editor, lesson)
    }

    tasksNumber = elements.filter { CommandFactory.getCommandType(it) == Command.CommandType.TRY }.count()
    currentTaskIndex = 0

    //Prepare environment before execution
    with(ExecutionList(myQueueOfElements, lesson, project, documentationMode)) {
      currentExecutionList = this
      CommandFactory.buildCommand(myQueueOfElements.peek(), documentationMode).execute(this)
    }
  }

  private fun createListOfCommands(root: Element): List<Element> {
    val commandsQueue = mutableListOf<Element>()
    for (rootChild in root.children) {
      //if element is MouseBlocked (blocks all mouse events) than add all children inside it.
      if (isMouseBlock(rootChild)) {
        if (rootChild.children != null) {
          commandsQueue.add(rootChild) //add block element
          for (mouseBlockChild in rootChild.children) {
            if (isCaretBlock(mouseBlockChild)) {
              if (mouseBlockChild.children != null) {
                commandsQueue.add(mouseBlockChild) //add block element
                commandsQueue += mouseBlockChild.children
                commandsQueue += Element(Command.CommandType.CARETUNBLOCK.toString()) //add unblock element
              }
            }
            else {
              commandsQueue.add(mouseBlockChild) //add inner elements
            }
          }
          commandsQueue.add(Element(Command.CommandType.MOUSEUNBLOCK.toString())) //add unblock element
        }
      }
      else if (isCaretBlock(rootChild)) {
        if (rootChild.children != null) {
          commandsQueue.add(rootChild) //add block element
          commandsQueue += rootChild.children
          commandsQueue += Element(Command.CommandType.CARETUNBLOCK.toString()) //add unblock element
        }
      }
      else {
        commandsQueue.add(rootChild)
      }
    }
    return commandsQueue
  }

  private fun isMouseBlock(el: Element): Boolean {
    return el.name.toUpperCase() == Command.CommandType.MOUSEBLOCK.toString()
  }

  private fun isCaretBlock(el: Element): Boolean {
    return el.name.toUpperCase() == Command.CommandType.CARETBLOCK.toString()
  }

}
